aip/0194.md

```diff
@@ -0,0 +1,84 @@
+---
+aip:
+  id: 194
+  state: reviewing
+  created: 2019-08-16
+permalink: /194
+redirect_from:
+  - /0194
+---
+
+# Retries for gRPC clients
+
+RPCs sometimes fail. When one does, the client performing the RPC needs to know
+whether it is safe to retry the operation. When status codes are used
+consistently across multiple APIs, clients can respond to failures
+appropriately.
+
+## Guidance
+
+Clients **should** retry requests for which repeated runs would not cause
+unintended state changes, which are non-transactional, and which are either
+unary or server streaming.
+
+Clients **should not** retry transactional requests; instead these requests
+**should** have user-level retry logic that retries the entire transaction
+block from the start.
+
+Clients **should not** retry requests in which repeated runs would cause
+unintended state changes.
+
+**Note:** This AIP does not cover client streaming or bi-directional streaming.
+
+### Retryable codes
+
+The following codes **should** be retried for retryable requests:
+
+| Code | Discussion |
+|--|--|
+| `INTERNAL` | This generally means that some internal part of the system has failed. These are retryable under the expectation that the system will recover from failures. |
```

Comment:
The "generally" bits here are quite worrying. For example, if an operation that would debit my bank account fails with internal or unavailable, I want to know *for sure* that it's not happened before I retry it.

If we have anything more concrete we can say to give us confidence in retrying here, that would be nice to see.

Reply:
Fixed (both "generally" cases (UNKNOWN, INTERNAL) are now marked "do not retry".